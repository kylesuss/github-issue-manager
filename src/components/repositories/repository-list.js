import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { Link } from 'react-router-dom'
import animateScrollTo from 'animated-scroll-to'
import { REPOSITORY_ISSUES_PATH } from 'constants/routes'
import githubData from 'components/containers/github-data'
import { GITHUB_REPOS_URL } from 'constants/github-api-routes'
import * as fonts from 'styles/fonts'
import * as spacing from 'styles/spacing'
import { buildRoute } from 'utils/routes'

const StyledRepositoryList = styled.div`
  display: flex;
  flex-direction: column;
`

const StyledLink = styled(Link)`
  color: #0366d6;
  text-decoration: none;
  font-weight: ${fonts.WEIGHT_SEMIBOLD};
  margin-bottom: ${spacing.HALF};
  
  &:hover {
    text-decoration: underline;
  }
`

const scrollToTop = () => animateScrollTo(0, {
  speed: 400
})

const RepositoryList = ({ data }) => {
  if (data.isLoading) {
    return <div>Loading</div>
  }

  if (data.error) {
    return <div>Error loading repos</div>
  }

  return (
    <StyledRepositoryList>
      {data.repos.map((repo) => {
        const linkTo = buildRoute(REPOSITORY_ISSUES_PATH, {
          owner: repo.owner.login,
          repo: repo.name
        })

        return (
          <StyledLink
            to={linkTo}
            key={repo.name}
            onClick={scrollToTop}
          >
            {repo.name}
          </StyledLink>
        )
      })}
    </StyledRepositoryList>
  )
}

RepositoryList.propTypes = {
  data: PropTypes.shape({
    error: PropTypes.string,
    isLoading: PropTypes.bool.isRequired,
    repos: PropTypes.arrayOf(PropTypes.shape({
      name: PropTypes.string.isRequired,
      owner: PropTypes.shape({
        login: PropTypes.string.isRequired
      }).isRequired
    }))
  })
}

const withGithubData = githubData(GITHUB_REPOS_URL, {
  name: 'repos'
})

export {
  RepositoryList
}

export default withGithubData(RepositoryList)
