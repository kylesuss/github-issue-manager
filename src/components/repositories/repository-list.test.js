import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { RepositoryList } from './repository-list'

test('it renders loading', () => {
  const wrapper = shallow(<RepositoryList data={{ isLoading: true }} />)
  expect(toJson(wrapper)).toMatchSnapshot()
})

test('it renders with an error', () => {
  const data = { isLoading: false, error: 'You broke it' }
  const wrapper = shallow(<RepositoryList data={data} />)
  expect(toJson(wrapper)).toMatchSnapshot()
})

test('it renders with data', () => {
  const data = {
    isLoading: false,
    response: [{ name: ':repoName1' }, { name: ':repoName2' }]
  }
  const wrapper = shallow(<RepositoryList data={data} />)
  expect(toJson(wrapper)).toMatchSnapshot()
})
