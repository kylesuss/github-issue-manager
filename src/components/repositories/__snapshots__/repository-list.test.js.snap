// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders loading 1`] = `
<div>
  Loading
</div>
`;

exports[`it renders with an error 1`] = `
<div>
  Error loading repos
</div>
`;

exports[`it renders with data 1`] = `
<repository-list__StyledRepositoryList>
  <repository-list__StyledLink
    key=":repoName1"
    onClick={[Function]}
    to="/repositories/:login1/:repoName1/issues"
  >
    :repoName1
  </repository-list__StyledLink>
  <repository-list__StyledLink
    key=":repoName2"
    onClick={[Function]}
    to="/repositories/:login2/:repoName2/issues"
  >
    :repoName2
  </repository-list__StyledLink>
</repository-list__StyledRepositoryList>
`;
